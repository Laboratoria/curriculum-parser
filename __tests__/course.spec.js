const mongoose = require('mongoose');
const models = require('@laboratoria/models')(mongoose);
const helpers = require('./helpers');
const course = require('../lib/course');


describe('course', () => {
  it('should reject with error when dir doesnt exist', () => (
    course('foo', models)
      .catch((err) => {
        expect(err.message).toMatch(/no such file or directory/);
        expect(err.code).toBe('ENOENT');
      })
  ));

  it('should reject when README.md is empty', () => (
    course(helpers.resolveFixturePath('00-course-empty'), models, {
      track: 'js',
      locale: 'es-ES',
    })
      .catch((err) => {
        expect(err.message).toBe('README.md del curso está vacío');
        expect(err.path).toBe(helpers.resolveFixtureDirReadmePath('00-course-empty'));
      })
  ));

  it('should reject when README.md doesnt start with h1', () => (
    course(helpers.resolveFixturePath('01-course-no-title'), models, {
      track: 'js',
      locale: 'es-ES',
    })
      .catch((err) => {
        expect(err.message).toBe('README.md del curso debe empezar con un h1 con el título del curso');
        expect(err.path).toBe(helpers.resolveFixtureDirReadmePath('01-course-no-title'));
      })
  ));

  it('should have empty tags if not found', () => (
    course(helpers.resolveFixturePath('02-course-no-tags'), models, {
      repo: 'Laboratoria/bootcamp',
      version: '2.0.0',
      track: 'js',
      locale: 'es-ES',
    })
      .then(data => expect(data.tags).toMatchSnapshot())
  ));

  it('should read primary (default) tags', () => (
    course(helpers.resolveFixturePath('02-course-tags'), models, {
      repo: 'Laboratoria/bootcamp',
      version: '2.0.0',
      track: 'js',
      locale: 'es-ES',
    })
      .then(data => expect(data.tags).toMatchSnapshot())
  ));

  it('should read main and secondary tags', () => (
    course(helpers.resolveFixturePath('02-course-secondary-tags'), models, {
      repo: 'Laboratoria/bootcamp',
      version: '2.0.0',
      track: 'js',
      locale: 'es-ES',
    })
      .then(data => expect(data.tags).toMatchSnapshot())
  ));

  it('should parse with target audience', () => (
    course(helpers.resolveFixturePath('02-course-with-target-audience'), models, {
      repo: 'Laboratoria/bootcamp',
      version: '2.0.0',
      track: 'js',
      locale: 'es-ES',
    })
      .then((data) => {
        expect(data.tags).toMatchSnapshot();
        expect(data.targetAudience).toMatchSnapshot();
      })
  ));

  it('should parse grades (evaluación) section', () => (
    course(helpers.resolveFixturePath('03-course-with-grades'), models, {
      repo: 'Laboratoria/bootcamp',
      version: '2.0.0',
      track: 'js',
      locale: 'es-ES',
    })
      .then(data => expect(data.grades).toMatchSnapshot())
  ));

  it('should trim <hr> from html fragments', () => (
    course(helpers.resolveFixturePath('03-course-with-grades'), models, {
      repo: 'Laboratoria/bootcamp',
      version: '2.0.0',
      track: 'js',
      locale: 'es-ES',
    })
      .then(data => expect(data.product).toMatchSnapshot())
  ));

  it('should validate units and parts', () => (
    course(helpers.resolveFixturePath('course-with-invalid-unit'), models, {
      repo: 'Laboratoria/bootcamp',
      version: '2.0.0',
      track: 'js',
      locale: 'es-ES',
    })
      .catch((err) => {
        expect(err.name).toBe('ValidationError');
        expect(err.message)
          .toBe('TopicUnitPart validation failed: type: Path `type` is required.');
      })
  ));

  it('should ignore tables in course description (taken from course readme)', () => (
    course(helpers.resolveFixturePath('00-course-with-part-tables'), models, {
      repo: 'Laboratoria/bootcamp',
      version: '2.0.0',
      track: 'js',
      locale: 'es-ES',
    })
      .then(result => expect(result.syllabus).toMatchSnapshot())
  ));
});
