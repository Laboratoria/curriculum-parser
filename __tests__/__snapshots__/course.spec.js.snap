// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`course should have empty tags if not found 1`] = `
Object {
  "primary": Object {},
  "secondary": Object {},
}
`;

exports[`course should ignore tables in course description (taken from course readme) 1`] = `
Object {
  "01-state": Object {
    "bonus": false,
    "description": "<p>En esta unidad veremos dos de los principios fundamentales de la programación
funcional: cómo evitar el estado compartido usando <strong>funciones puras</strong> y el
concepto de <strong>inmutabilidad</strong>.</p>
",
    "order": 0,
    "parts": Object {
      "00-opening": Object {
        "body": "<h2 id=\\"¿por-qué-aprender-esto\\">¿Por qué aprender esto?</h2>
<p>En esta unidad veremos dos de los principios fundamentales de la programación
funcional: cómo evitar el estado compartido usando <strong>funciones puras</strong> y el
concepto de <strong>inmutabilidad</strong>. Ambos conceptos son esenciales en el paradigma
funcional y su uso presenta un montón de beneficios, a veces sutiles, pero muy
importantes.</p>
<ul>
<li>Representan bloques de código reusable dentro de un programa.</li>
<li>Promueven buenas prácticas de desarrollo como [DRY][]
(<em>Don’t Repeat Yourself</em>) y [KISS][] (<em>Keep It Simple, Stupid</em>).</li>
<li>Al no depender del contexto externo, son inmunes a toda clase de errores o
<em>bugs</em> que tienen que ver con el estado mutable compartido.</li>
<li>Su naturaleza independiente las hace excelentes candidatos para procesamiento
concurrente a lo largo de muchos CPUs e incluso para la computación
distribuida, lo cual las hace esenciales para la ejecución de tareas de
cálculo científico y de uso intensivo de recursos.</li>
<li>Su aislamiento facilita la refactorización y reorganización del código,
haciendo tú código más adaptable y flexible a futuros cambios.</li>
<li>Es mucho más sencillo el desarrollo de pruebas unitarias contra funciones
puras.</li>
<li>Las funciones puras representan la base de la programación funcional.</li>
</ul>
<h2 id=\\"guía-de-preguntas-y-conceptos-clave\\">Guía de preguntas y conceptos clave</h2>
<p>Cuando empezamos a estudiar un tema nuevo, es útil tener una idea de los
conceptos más importantes de lo que vamos a aprender y de los temas centrales
que debemos prestar particular atención.</p>
<p>A continuación te presentamos una serie de preguntas que debes ser capaz de
responder al terminar esta unidad. Utiliza estas preguntas como guía para
orientar tus esfuerzos de aprendizaje. Regresa a ellas constantemente a medida
que avanzas para validar que estás avanzando en la dirección correcta. Que te
sirva como un \\"checklist\\" que vas marcando a medida que vas progresando.</p>
<ul>
<li>Cómo evitar que nuestras funciones tengan efectos colaterales?</li>
<li>Por qué es beneficioso evitar el estado compartido?</li>
<li>Cuáles son las ventajas que ofrecen las funciones puras, en particular a la
hora de hacer testing?</li>
<li>Cómo evitar la mutación de objetos y arreglos?</li>
<li>Qué tipos de datos son mutables e inmutables en JavaScript?</li>
<li>Cómo usar métodos de arreglos como <code>Array#concat</code>, <code>Array#slice</code>, <code>Array#map</code>,
<code>Array#filter</code>, ... para crear arreglos nuevos en vez de mutar (modificar)
valores existentes?</li>
<li>Cómo usar métodos de objetos (por ejemplo <code>Object.assign()</code>) para copiar
propiedades y así <em>clonar</em> objetos?</li>
<li>Cómo usar el <em>spread operator</em> (<code>...</code>) para copiar objetos y arreglos?</li>
</ul>
",
        "duration": 5,
        "durationString": "5min",
        "format": "self-paced",
        "order": 0,
        "title": "Apertura",
        "type": "read",
      },
      "01-pure-fun": Object {
        "body": "<p>Las funciones puras tienen aplicaciones en muchos ámbitos, entre ellos la
programación funcional. También facilitan la concurrencia, y como veremos en
siguientes etapas serán muy empleadas en aplicaciones basadas en React+Redux.</p>
<p>Anteriormente mencionamos que una función es básicamente un proceso que toma una
entrada o argumentos, y produce una salida o valor de retorno. También hemos
visto que las funciones se emplean con ciertos propósitos:</p>
<ul>
<li><em>Manipulación de datos:</em> Transforma una serie de argumentos o entrada en un
valor de retorno. Un ejemplo de esto sería hacer <em>flattening</em><sup>1</sup> de
un objeto muy anidado.</li>
<li><em>Procedimientos:</em> Una función puede ser llamada para realizar una serie de
pasos (receta). <strong>La secuencia es conocida como procedimiento</strong> (primero haz
esto, luego haz aquello), el estilo de programación bajo este estilo se
denomina <strong>programación por procedimientos</strong>, la cual está enmarcada en la
<strong>programación imperativa</strong>, donde se usan sentencias que modifican o mutan el
estado del programa, de la misma forma en la que mandatos imperativos expresan
comandos en nuestro lenguaje natural, un programa imperativo consiste en
comandos que la computadora debe realizar. La programación imperativa se
enfoca en describir <em>cómo</em> un programa opera, mientras que en la programación
funcional nos enfocamos en <em>qué</em> debe realizar un programa sin especificar
<em>cómo</em> dicho programa debe alcanzar el resultado.</li>
<li><em>Entrada/Salida:</em> Algunas funciones existen para comunicarse con otras partes
o subsistemas, por ejemplo: la pantalla, almacenamiento, registro de
operaciones en disco, operaciones a través de la red.</li>
</ul>
<p>Una vez vistas las diversas aplicaciones de las funciones procedamos a dejar
claro que significa una <em>función pura</em>.</p>
<blockquote>
<p>Toda función que dados los mismos inputs siempre retorna lo mismo, y sin
efectos secundarios, es una <em>función pura</em>.</p>
</blockquote>
<p>En <strong>programación funcional</strong>, el comportamiento de las funciones depende de una
y solo una cosa: los argumentos pasados explícitamente a la función. Esto quiere
decir que si proporcionas los mismos datos como argumentos o entrada, la función
siempre debe producir el mismo valor de retorno. A esta propiedad se le conoce
como <strong>transparencia referencial</strong>.</p>
<p>Lo mencionado en el párrafo anterior, hace que en las funciones puras sea más
fácil apreciar todas las circunstancias que pueden presentarse, incluyendo
aquellos escenarios que resultan en errores. Escribir funciones que solo
dependen de sus argumentos para definir su comportamiento también facilita
replicar <em>bugs</em> o poner en práctica <em>Test-Driven Development</em> (TDD por sus
siglas en inglés).</p>
<p>Comencemos desarrollando nuestros casos de pruebas:</p>
<p><strong>lowercaser.test.js.</strong></p>
<pre><code class=\\"language-js\\">const lowerCaser = require('./lowercaser')

test('Should take an input string and returns it lowercased', () =&gt; {
  expect(lowerCaser('LOREM IPSUM')).toBe('lorem ipsum')
})
</code></pre>
<hr>
<p>NOTA:</p>
<p>A lo largo del curso usaremos <a href=\\"https://facebook.github.io/jest/\\" target=\\"_blank\\">Jest</a> como <em>framework</em> para la ejecución de
pruebas unitarias. Puedes instalar <strong>Jest</strong> por medio de <code>npm</code> al ejecutar el
siguiente comando:</p>
<pre><code class=\\"language-sh\\">npm install --save-dev jest
</code></pre>
<p>O también puedes usar <code>yarn</code> al ejecutar:</p>
<pre><code class=\\"language-sh\\">yarn add --dev jest
</code></pre>
<p>Para correr las pruebas haremos:</p>
<pre><code class=\\"language-sh\\">yarn run jest -- fichero.test.js
</code></pre>
<hr>
<p>Ahora, pasemos a implementar lo especificado en nuestros casos de pruebas:</p>
<p><strong>lowercaser.js.</strong></p>
<pre><code class=\\"language-js\\">const lowerCaser = input =&gt; input.toString().toLowerCase()

module.exports = lowerCaser
</code></pre>
<p>Una vez completada nuestra implementación inicial, corroboremos nuestro trabajo
por medio de la ejecución de las pruebas:</p>
<pre><code class=\\"language-sh\\">$ yarn run jest -- lowercaser.test.js

PASS  ./lowercaser.test.js
   ✓ Should take an input string and returns it lowercased (15ms)

   Test Suites: 1 passed, 1 total
   Tests:       1 passed, 1 total
   Snapshots:   0 total
   Time:        0.859s, estimated 1s
   Ran all test suites matching \\"lowercaser.test.js\\".
</code></pre>
<p>Parece que todo funciona como se espera. Continuemos.</p>
<h2 id=\\"dada-la-misma-entrada-devuelve-siempre-la-misma-salida\\">Dada la misma entrada, devuelve siempre la misma salida</h2>
<p>Con nuestra función <code>lowerCaser()</code>, podemos reemplazar la llamada de la función
por el resultado, y el código tendrá el mismo significado <code>lowerCaser('LOREM IPSUM')</code> siempre será lo mismo que <code>lorem ipsum</code> en su programa, sin importar el
contexto, cuántas veces lo llame o cuándo lo llame.</p>
<p>Pero no se puede decir lo mismo de todas las funciones. Algunas funciones se
basan en información distinta de los argumentos que se transmiten para producir
resultados. Considera este ejemplo:</p>
<pre><code class=\\"language-js\\">Math.random() // =&gt; 0.4011148700956255
Math.random() // =&gt; 0.8533405303023756
Math.random() // =&gt; 0.3550692005082965
</code></pre>
<p>A pesar de que no pasamos ningún argumento en ninguna de las llamadas a la
función, todos produjeron resultados diferentes, lo que significa que
<code>Math.random()</code> <strong>no es una función pura</strong>. <code>Math.random()</code> produce un nuevo
número aleatorio entre 0 y 1 cada vez que lo ejecutas, entonces es obvio que no
se podría simplemente reemplazarlo por <code>0.4011148700956255</code> sin cambiar el
significado del programa.</p>
<p>Eso produciría el mismo resultado cada vez que se ejecute el programa. Cuando le
pedimos a la computadora un número aleatorio, por lo general significa que
queremos un resultado diferente al que obtuvimos la última vez. ¿Cuál es el
sentido de un par de dados con los mismos números impresos en todas las caras? A
veces tenemos que preguntarle a la computadora por la hora actual. No vamos a
entrar en detalles de cómo funcionan las funciones de tiempo. Por ahora,
simplemente copia este código:</p>
<pre><code class=\\"language-js\\">const time = () =&gt; new Date().toLocaleTimeString()
time() // =&gt; \\"5:15:45 PM\\"
</code></pre>
<p>¿Qué sucedería si reemplazo la llamada de la función <code>time()</code> con la hora
actual? Siempre diría que es la misma hora: la hora con la cual la llamada a la
función se reemplazó. En otras palabras, solo podría producir la salida correcta
una vez al día, y solo si se ejecuta el programa en el momento exacto en que la
hora se reemplazó por la función.</p>
<p>Entonces, claramente, <code>time()</code> no es como la función <code>lowerCaser()</code>.</p>
<p>Una función solo es pura si, dada la misma entrada, siempre producirá la misma
salida. Tal vez recuerdes esta regla de la clase de álgebra: los mismos valores
de entrada se asignarán siempre al mismo valor de salida. Sin embargo, muchos
valores de entrada se pueden asignar al mismo valor de salida. Por ejemplo, la
siguiente función es pura:</p>
<pre><code class=\\"language-js\\">const highpass = (cutoff, value) =&gt; value &gt;= cutoff;
</code></pre>
<p>Los mismos valores de entrada se asignarán siempre al mismo valor de salida:</p>
<pre><code class=\\"language-js\\">highpass(5, 5) // =&gt; true
highpass(5, 5) // =&gt; true
highpass(5, 5) // =&gt; true
</code></pre>
<p>Muchos valores de entrada tal vez se pueden asignar al mismo valor de salida:</p>
<pre><code class=\\"language-js\\">highpass(5, 123) // true
highpass(5, 6)   // true
highpass(5, 18)  // true
highpass(5, 1)   // false
highpass(5, 3)   // false
highpass(5, 4)   // false
</code></pre>
<h2 id=\\"efectos-colaterales\\">Efectos colaterales</h2>
<p>En este punto cabe aclarar que las funciones puras no producen <strong>efectos
colaterales</strong> o <em>side-effects</em>, pues el propósito de vida de una función pura es
únicamente calcular el valor de retorno, solo eso y nada más.</p>
<p>En las ciencias de la computación, una función o expresión se dice que tiene
efectos colaterales si modifica algún estado fuera de su ámbito (<em>scope</em>), o si
tiene interacciones observables con la función que la invocó, o si modifica el
mundo exterior aparte de retornar el valor esperado. Por ejemplo, una función
particular podría modificar una variable global (estado fuera de su ámbito),
modificar uno de los argumentos mutables (interactuar con la función la invocó),
levantar una excepción, escribir datos a un fichero o llamar a otra función que
si tiene efectos colaterales. Ten en cuenta que ante la presencia de funciones
con efectos colaterales, el comportamiento de un programa podría depender de su
historia, esto quiere decir que el orden de evaluación importa. Por lo tanto, la
comprensión y análisis de una función con efectos colaterales requiere
conocimientos acerca de su contexto y su posible historia, lo cual hace más
difícil la corrección de errores. En la programación funcional, los efectos
colaterales se usan con moderación.</p>
<p>Veamos un ejemplo de efecto colateral y cómo podemos evitarlo.</p>
<p>Supongamos que deseamos crear una función que concatene dos <em>arrays</em> que pueden
contener elementos anidados, tratemos de emular el siguiente comportamiento:</p>
<p><strong>concat.test.js.</strong></p>
<pre><code class=\\"language-js\\">const concat = require('./concat')

test('should concatenate array on the right with array on the left', () =&gt; {
  const left = [1, [2], 3]
  const right = [[4], 5, [6, 7], 8]
  const expected = [1, [2], 3, [4], 5, [6, 7], 8]

  expect(concat(left, right)).toEqual(expected)
})
</code></pre>
<p>Una implementación inicial puede ser la siguiente:</p>
<p><strong>concat.js.</strong></p>
<pre><code class=\\"language-js\\">const concat = (left, right) =&gt; {
  const result = left

  right.map(item =&gt; {
    result.push(item)
  })

  return result
}

module.exports = concat
</code></pre>
<blockquote>
<p><strong>Nota</strong></p>
<p>Ten en cuenta que esta implementación es a modo de ejemplo, en realidad
podríamos usar <a href=\\"https://developer.mozilla.org/es/docs/Web/JavaScript/Referencia/Objetos_globales/Array/concat\\" target=\\"_blank\\">Array.prototype.concat()</a> directamente.</p>
</blockquote>
<p>Comprobemos nuestro trabajo:</p>
<pre><code class=\\"language-sh\\">$ yarn run jest -- concat.test.js

 PASS  ./concat.test.js
   ✓ should concatenate the array on the right with the array on the left (3ms)

   Test Suites: 1 passed, 1 total
   Tests:       1 passed, 1 total
   Snapshots:   0 total
   Time:        0.67s, estimated 1s
   Ran all test suites matching \\"concat.test.js\\".
</code></pre>
<p>Todo parece indicar que nuestra función tiene un comportamiento correcto, ahora
agreguemos un caso de prueba para comprobar que no estamos alterando el contexto
externo:</p>
<p><strong>concat.test.js.</strong></p>
<pre><code class=\\"language-js\\">const concat = require('./concat')

test('should concatenate array on the right with array on the left', () =&gt; {
  const left = [1, [2], 3]
  const right = [[4], 5, [6, 7], 8]
  const expected = [1, [2], 3, [4], 5, [6, 7], 8]

  expect(concat(left, right)).toEqual(expected)
})

test('should not mutate external context', () =&gt; {
  const left = [1, [2], 3]
  const right = [[4], 5, [6, 7], 8]
  const backup = Object.assign([], left)

  concat(left, right)

  expect(left).toEqual(backup)
})
</code></pre>
<p>Al ejecutar las pruebas obtenemos lo siguiente:</p>
<pre><code class=\\"language-sh\\">$ yarn run jest -- concat.test.js
 FAIL  ./concat.test.js
  ● should not mutate external context

    expect(received).toEqual(expected)

    Expected value to equal:
      [1, [2], 3]
    Received:
      [1, [2], 3, [4], 5, [6, 7], 8]

  ...

  ✓ should concatenate the array on the right with the array on the left (4ms)
  ✕ should not mutate external context (6ms)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 passed, 2 total
Snapshots:   0 total
Time:        1.011s
Ran all test suites matching \\"concat.test.js\\".
</code></pre>
<p><strong>Ouch!</strong>, con esto se demuestra que nuestra <strong>función no es pura</strong>, pues
estamos alterando el contexto externo, lo cual es un tipo de efecto colateral.</p>
<p>El comportamiento anterior sucede porque los objetos o arreglos pasados a las
funciones como argumento se pasan por referencia, no por copia, lo que significa
que si una función muta una propiedad en un objeto o arreglo, supondría que
dicha mutación sería accesible desde afuera de la función. Las funciones puras
no deben alterar el estado externo.</p>
<p>Si bien el valor de retorno de nuestra función es el esperado, el problema con
la implementación actual es que hemos mutado un estado compartido. Imagina por
un momento que otras funciones pueden depender del estado del arreglo u objeto
asumiendo que su estado es el original (antes de llamar a nuestra función
<code>concat</code>), y ahora que hemos mutado ese estado compartido, tenemos que
preocuparnos por el impacto que tendrá dicho cambio en la lógica del programa si
cambiamos el orden en que se han llamado las funciones. Refactorizar el código
podría resultar en errores apareciendo aquí y allá, lo que podría arruinar la
lógica general de nuestra aplicación, y como resultado muchos clientes
disgustados.</p>
<p>Veamos ahora cómo podemos corregir esta situación:</p>
<p><strong>concat.js.</strong></p>
<pre><code class=\\"language-js\\">const concat = (left, right) =&gt; {
  const result = Object.assign([], left)

  right.map(item =&gt; {
    result.push(item)
  })

  return result
}

module.exports = concat
</code></pre>
<p>Ahora al ejecutar las pruebas obtenemos lo siguiente:</p>
<pre><code class=\\"language-sh\\">$ yarn run jest -- concat.test.js
 PASS  ./concat.test.js
  ✓ should concatenate the array on the right with the array on the left (5ms)
  ✓ should not mutate external context (1ms)

Test Suites: 1 passed, 1 total
Tests:       2 passed, 2 total
Snapshots:   0 total
Time:        0.977s
Ran all test suites matching \\"concat.test.js\\".
</code></pre>
<p>¿Recuerdas que antes habíamos mencionado que podíamos haber usado
<code>Array.prototype.concat</code> directamente?, pues veamos una simplificación del
código:</p>
<p><strong>concat.js.</strong></p>
<pre><code class=\\"language-js\\">const concat = (left, right) =&gt; {
  return left.concat(right)
}

module.exports = concat
</code></pre>
<p>Si ejecutas de nuevo las pruebas unitarias verás que cumplimos con todas las
condiciones.</p>
<hr>
<p>NOTA: Es común usar métodos como <code>Array#slice</code>, <code>Array#map</code> o <code>Array#filter</code>
además de <code>Array#concat</code> para crear copias de arreglos. En <strong>ES6</strong> también
tenemos el <em>spread operator</em> que nos permite copiar las propiedades enumerables
de un arreglo u objeto de esta forma:</p>
<pre><code class=\\"language-js\\">const array = [1, 2, 3];
const arrayCopy = [...a]; // \`b\` es un nuevo arreglo con los elementos de \`a\`.

const obj = { foo: 'bar' };
const objCopy = {...obj};
</code></pre>
<hr>
<h2 id=\\"beneficios\\">Beneficios</h2>
<p>Una vez analizadas las funciones puras, volvamos a repasar los beneficios que
ofrecen, tal como vimos en la apertura de esta unidad:</p>
<ul>
<li>Toman ciertos argumentos como entrada y generan un valor de retorno que
exclusivamente depende de los argumentos dados.</li>
<li>Representan bloques de código reusable dentro de un programa.</li>
<li>Promueven buenas prácticas de desarrollo como <a href=\\"https://en.wikipedia.org/wiki/Don%27t_repeat_yourself\\" target=\\"_blank\\">DRY</a>
(<em>Don’t Repeat Yourself</em>) y <a href=\\"https://en.wikipedia.org/wiki/KISS_principle\\" target=\\"_blank\\">KISS</a> (<em>Keep It Simple, Stupid</em>).</li>
<li>Al no depender del contexto externo, son inmunes a toda clase de errores o
<em>bugs</em> que tienen que ver con el estado mutable compartido.</li>
<li>Su naturaleza independiente las hace excelentes candidatos para procesamiento
concurrente a lo largo de muchos CPUs e incluso para la computación
distribuida, lo cual las hace esenciales para la ejecución de tareas de
cálculo científico y de uso intensivo de recursos.</li>
<li>Su aislamiento facilita la refactorización y reorganización del código,
haciendo tú código más adaptable y flexible a futuros cambios.</li>
<li>Es mucho más sencillo el desarrollo de pruebas unitarias contra funciones
puras.</li>
<li>Las funciones puras representan la base de la programación funcional.</li>
</ul>
<p>Por las razones antes mencionadas, recomendamos favorecer la implementación de
funciones puras. Por lo tanto, siempre que sea práctico implementar los
requerimientos de un programa usando funciones puras, debes darle preferencia
sobre otras opciones.</p>
<h2 id=\\"referencias\\">Referencias</h2>
<ul>
<li>Conrad Barski. 2010. Land of Lisp: Learn to Program in Lisp, One Game at a
Time! (1st ed.). No Starch Press, San Francisco, CA, USA. Page 301.</li>
<li><a href=\\"https://medium.com/javascript-scene/master-the-javascript-interview-what-is-a-pure-function-d1c076bec976\\" target=\\"_blank\\">Master the JavaScript Interview: What is a Pure Function?</a>. En Medium.
Consultado el 11 de Julio de 2017</li>
<li><a href=\\"https://en.wikipedia.org/wiki/Side_effect_(computer_science)\\" target=\\"_blank\\">Side effect (computer science)</a></li>
<li><a href=\\"https://en.wikipedia.org/wiki/Imperative_programming\\" target=\\"_blank\\">Imperative Programming</a></li>
<li><a href=\\"https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLowerCase\\" target=\\"_blank\\">String.prototype.toLowerCase()</a></li>
<li><a href=\\"https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random\\" target=\\"_blank\\">Math.random()</a></li>
<li><a href=\\"https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString\\" target=\\"_blank\\">Date.prototype.toLocaleDateString()</a></li>
<li><a href=\\"https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat\\" target=\\"_blank\\">Array.prototype.concat()</a></li>
<li><a href=\\"https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign\\" target=\\"_blank\\">Object.assign()</a></li>
</ul>
<h2 id=\\"notas-al-pie\\">Notas al pie</h2>
<ol>
<li>Algunas veces necesitamos aplanar árboles u objetos muy anidados que son
resultado de una consulta o <em>query</em>, un patrón común es convertirlos en
arreglos o <em>arrays</em> para luego poder aplicar operaciones como <code>filter()</code> o
<code>map()</code> sobre ellos.</li>
</ol>
",
        "duration": 30,
        "durationString": "30min",
        "format": "self-paced",
        "order": 1,
        "title": "Funciones puras",
        "type": "read",
      },
    },
    "stats": Object {
      "duration": 35,
      "durationString": "35min",
      "exerciseCount": 0,
      "partCount": 2,
    },
    "title": "Evitando el estado compartido y cambiante",
  },
}
`;

exports[`course should parse grades (evaluación) section 1`] = `
"<ul>
<li><strong>Ejercicios</strong>: Durante el curso completarás varios ejercicios en el LMS.
Estos ejercicios incluyen tests automatizados con la intención de dar feedback
immediato, pero los ejercicios en sí no serán calificados. Se elegirá uno o
dos ejercicios junto con el proyecto para la sesión de <em>code review</em>, que sí
será calificada.</li>
<li><strong>Cuestionarios</strong>: Al igual que los ejercicios, a lo largo del curso
responderás varios cuestionarios con feedback immediato, y éstos no cuentan
para la nota final.</li>
<li><strong>Proyecto</strong>: Antes de la sesión de <em>code review</em> y las demos, cada alumna
debe entregar (via pull request) el código de su proyecto. El proyecto incluye
refactorizar varios módulos de una base de código existente.</li>
<li><strong>Code review</strong>: <strong>(50%)</strong> La última semana del curso tendrás una sesión de
<em>code review</em> con uno de lxs instructorxs. En esta sesión se revisará el
código del proyecto así como uno o dos ejercicios realizados durante el curso.</li>
<li><strong>Demo</strong>: <strong>(50%)</strong> El curso cerrará con una demo en la que tendrás que
presentar al resto de la clase el trabajo realizado y lo aprendido en el
proyecto.</li>
</ul>
"
`;

exports[`course should parse with target audience 1`] = `
Object {
  "primary": Object {
    "bar": true,
    "baz": true,
    "foo": true,
  },
  "secondary": Object {},
}
`;

exports[`course should parse with target audience 2`] = `
"<p>Este curso está dirigido a desarrolladoras tanto <em>front-end</em> como <em>back-end</em>.
JavaScript es un lenguaje de naturaleza funcional. En este curso aprenderás
sobre los principios de la programación funcional y cómo se reflejan en
JavaScript moderno.</p>
"
`;

exports[`course should read main and secondary tags 1`] = `
Object {
  "primary": Object {
    "bar": true,
    "baz": true,
    "foo": true,
  },
  "secondary": Object {
    "one": true,
    "three": true,
    "two": true,
  },
}
`;

exports[`course should read primary (default) tags 1`] = `
Object {
  "primary": Object {
    "bar": true,
    "baz": true,
    "foo": true,
  },
  "secondary": Object {},
}
`;

exports[`course should trim <hr> from html fragments 1`] = `
"<ul>
<li>Refactorizar juego usado en lección 9 usando principios de programación
funcional, ES6 y añadiendo un conjunto de pruebas.</li>
<li>El &quot;proyecto&quot; está basado en iterar una aplicación existente y llevar a cabo
su siguiente release usando metodología ágil.</li>
</ul>
"
`;
